tss-esapi/src/abstraction/cipher.rs

```diff
@@ -0,0 +1,241 @@
+// Copyright 2019 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use crate::{
+    constants::AlgorithmIdentifier,
+    interface_types::{
+        algorithm::{HashingAlgorithm, SymmetricAlgorithm, SymmetricMode, SymmetricObject},
+        key_bits::{AesKeyBits, CamelliaKeyBits, Sm4KeyBits},
+    },
+    structures::{SymmetricCipherParameters, SymmetricDefinition, SymmetricDefinitionObject},
+    Error, Result, WrapperErrorKind,
+};
+use std::convert::{TryFrom, TryInto};
+/// Block cipher identifiers
+///
+/// Structure useful for handling an abstract representation of ciphers. Ciphers are
+/// defined foremost through their symmetric algorithm and, depending on the type of that
+/// algorithm, on a set of other values.
+#[derive(Copy, Clone, Debug)]
+pub struct Cipher {
+    algorithm: SymmetricAlgorithm,
+    mode: Option<SymmetricMode>,
+    key_bits: Option<u16>,
+    hash: Option<HashingAlgorithm>,
+}
```

Comment:
Does it make any sense to keep this if we have the `SymmetricDefinition` and `SymmetricDefinitionObject` structs?

Reply:
I think yes. Because it is one type that can generate them both so that you avoid specify the same thing twice which can be nice.